internal/repoconfig/config.go

```diff
@@ -0,0 +1,187 @@
+type TestFile struct {
+	Path   string            // Full path relative to repository root
+	Labels map[string]string // Labels from all parent directories
+}
+
+// generateRunID creates a random alphanumeric string for the run
+func generateRunID() string {
+	const charset = "abcdefghijklmnopqrstuvwxyz0123456789"
+	const length = 8
+
+	// Initialize random seed
+	rand.Seed(time.Now().UnixNano())
+
+	// Generate random string
+	b := make([]byte, length)
+	for i := range b {
+		b[i] = charset[rand.Intn(len(charset))]
+	}
+	return string(b)
+}
+
+// findGitRepoRoot finds the root of the git repository using go-git
+func findGitRepoRoot(startPath string) (string, error) {
+	repo, err := git.PlainOpenWithOptions(startPath, &git.PlainOpenOptions{
+		DetectDotGit: true,
+	})
+	if err != nil {
+		return "", fmt.Errorf("not a git repository (or any parent up to mount point %s): %w", startPath, err)
+	}
+
+	worktree, err := repo.Worktree()
+	if err != nil {
+		return "", fmt.Errorf("failed to get worktree: %w", err)
+	}
+
+	return worktree.Filesystem.Root(), nil
+}
+
+// LoadConfig reads the .signadot/config.yaml file from the git repository root
+func LoadConfig(startPath string) (*Config, error) {
+	// Find git repository root
+	repoRoot, err := findGitRepoRoot(startPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find git repository root: %w", err)
+	}
+
+	configPath := filepath.Join(repoRoot, ".signadot", "config.yaml")
+	fmt.Printf("Reading config from: %s\n", configPath)
+	data, err := os.ReadFile(configPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read .signadot/config.yaml: %w", err)
+	}
+
+	var cfg Config
+	if err := yaml.Unmarshal(data, &cfg); err != nil {
+		return nil, fmt.Errorf("failed to parse .signadot/config.yaml: %w", err)
+	}
+
+	if len(cfg.SmartTests) == 0 {
+		return nil, fmt.Errorf("smart_tests is required in .signadot/config.yaml")
+	}
+
+	return &cfg, nil
+}
+
+// sortPathsByLength sorts paths by their length (shortest first)
+func sortPathsByLength(paths []string) []string {
+	sorted := make([]string, len(paths))
+	copy(sorted, paths)
+	sort.Slice(sorted, func(i, j int) bool {
+		return len(sorted[i]) < len(sorted[j])
+	})
+	return sorted
+}
+
+// FindTestFiles finds all test files in the tests directories
+func FindTestFiles(startPath string, cfg *Config) ([]TestFile, error) {
+	// Find git repository root
+	repoRoot, err := findGitRepoRoot(startPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find git repository root: %w", err)
+	}
+
+	// Generate a run ID for this execution
+	runID := generateRunID()
+
+	// Build the directory labels cache
+	dirLabelsCache, err := buildLabelsCache(repoRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	// Use a map to deduplicate files by their full path
+	testFileMap := make(map[string]TestFile)
+
+	// Sort paths by length to process parent directories first
+	sortedPaths := sortPathsByLength(cfg.SmartTests)
```

Comment:
Probably best to clean the paths from the config file to avoid noise like ../dir/.. and ././ etc